### PR TITLE
Pclouds 2649/adjust autodiscovery mechanism for multi project approach

### PR DESCRIPTION
**Pull request by szymon-nagel-dt**
Added multiproject support

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -189,32 +189,35 @@ async def fetch_ingest_lines_task(context: MetricsContext, project_id: str, serv
...
+                    context=context,
+                    project_id=project_id,
+                    service=service,
+                    metric=metric
+                )
+                fetch_metric_coros.append(fetch_metric_coro)
+
+                if metric.autodiscovered_metric:
+                    metrics_metadata.append(MetadataIngestLine(
+                        metric_name = metric.dynatrace_name,
```
Comment (nikita-paskal-dynatrace) [inline src/main.py:215]: https://peps.python.org/pep-0008/#whitespace-in-expressions-and-statements

File: src/lib/autodiscovery/gcp_metrics_descriptor.py [hunk 2]
```diff
@@ -85,10 +87,10 @@ def _cast_metric_kind_to_dt_format(metric_kind: str, value_type: str) -> str:
         elif metric_kind == "CUMULATIVE":
             return "count,delta"
         else:
-            raise Exception("Unknown metric type")
+            raise Exception(f"Unknown metric type {metric_kind}")
```
Comment (nikita-paskal-dynatrace) [inline src/lib/autodiscovery/gcp_metrics_descriptor.py:90]: raise Exception(f"Unknown metric type`:` {metric_kind}")

File: src/lib/autodiscovery/gcp_metrics_descriptor.py [hunk 3]
```diff
@@ -85,10 +87,10 @@ def _cast_metric_kind_to_dt_format(metric_kind: str, value_type: str) -> str:
         elif metric_kind == "CUMULATIVE":
             return "count,delta"
         else:
```
Comment (nikita-paskal-dynatrace) [inline src/lib/autodiscovery/gcp_metrics_descriptor.py:89]: As I told before: no need for this check. You can only keep if/elif and raise exception at the end

File: src/lib/autodiscovery/gcp_metrics_descriptor.py [hunk 4]
```diff
@@ -85,10 +87,10 @@ def _cast_metric_kind_to_dt_format(metric_kind: str, value_type: str) -> str:
...
-            raise Exception("Unknown metric type")
+            raise Exception(f"Unknown metric type {metric_kind}")
 
     @staticmethod
-    def _get_key_metric_sufix(metric_name: str, data_type: str) -> str:
+    def _get_key_metric_suffix(metric_name: str, data_type: str) -> str:
         if metric_name.endswith(("_count", ".count")) and data_type == "gauge":
             return ".gauge"
         if (
```
Comment (nikita-paskal-dynatrace) [inline src/lib/autodiscovery/gcp_metrics_descriptor.py:96]: metric_name.endswith(("_count", ".count" ))

File: src/lib/autodiscovery/gcp_metrics_descriptor.py (hunk 3 above)
Comment (szymon-nagel-dt) [reply src/lib/autodiscovery/gcp_metrics_descriptor.py:89]: I thing both ways are good: https://peps.python.org/pep-0008/#programming-recommendations
image

In this codebase it's usually to have with else approach.
Even with Your code 😄:
https://github.com/dynatrace-oss/dynatrace-gcp-monitor/commit/7c0952d9cb37cce30500c9523574aff84306a2d7#diff-9f71d84264b0b7aae201479a9cb46ca9edfd78dea0b37687ffcbcf28f070c58fR64
I duno what version in "better" 😉

File: src/lib/autodiscovery/autodiscovery.py [hunk 5]
```diff
@@ -1,69 +1,140 @@
...
+        print_metric_ingest_input=config.print_metric_ingest_input(),
+        self_monitoring_enabled=config.self_monitoring_enabled(),
+        scheduled_execution_id=logging_context.scheduled_execution_id,
+    )
+
+    projects_ids = await get_all_accessible_projects(context, gcp_session, token)
+    disabled_projects = []
+
+    if not config.scoping_project_support_enabled():
+        disabled_project, _ = await get_disabled_projects_and_disabled_apis_by_project_id(
```
[the remark this answers is not in the document]
Comment (szymon-nagel-dt) [reply src/lib/autodiscovery/autodiscovery.py:65]: This is not even my function, i am just using it 🤣

**Review by nikita-paskal-dynatrace — APPROVED** [review]

**Pull request merged**